# Исследование структуры DataFrame


## Вывод первых и последних строк

В большинстве случаев для того, чтобы понять структуру DataFrame и удостовериться, что таблица подгрузилась верно, достаточно вывести несколько первых или последних строк.

Для этого у DataFrame есть методы head() и tail(), которые возвращают n первых и n последних строк таблицы соответственно (по умолчанию n = 5).

In [2]:
# Импортируем библиотеку и DataFrame
import pandas as pd
melb_data = pd.read_csv('data/melb_data.csv', sep=',')

# Выведем первые пять строк нашей таблицы:
display(melb_data.head())

index      Suburb           Address  Rooms Type      Price Method SellerG  \
0      0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1      1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2      2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3      3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4      4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  ...  Car  Landsize  BuildingArea  YearBuilt  \
0  3/12/2016       2.5  ...  1.0     202.0         126.0     1970.0   
1  4/02/2016       2.5  ...  0.0     156.0          79.0     1900.0   
2  4/03/2017       2.5  ...  0.0     134.0         150.0     1900.0   
3  4/03/2017       2.5  ...  1.0      94.0         126.0     1970.0   
4  4/06/2016       2.5  ...  2.0     120.0         142.0     2014.0   

   CouncilArea  Lattitude  Longtitude             Regionname  Propertycount  \
0        Yarra   -37.7996    144.9984  Northern Metropolitan         4019.0   
1        Yarra   -37.8079    144.9934  Northern Metropolitan         4019.0   
2        Yarra   -37.8093    144.9944  Northern Metropolitan         4019.0   
3        Yarra   -37.7969    144.9969  Northern Metropolitan         4019.0   
4        Yarra   -37.8072    144.9941  Northern Metropolitan         4019.0   

          Coordinates  
0  -37.7996, 144.9984  
1  -37.8079, 144.9934  
2  -37.8093, 144.9944  
3  -37.7969, 144.9969  
4  -37.8072, 144.9941  

[5 rows x 23 columns]

In [6]:
# Код ниже выводит семь последних строк нашей таблицы:
melb_data.tail(7)

index         Suburb        Address  Rooms Type      Price Method  \
13573  13573       Werribee    5 Nuragi Ct      4    h   635000.0      S   
13574  13574    Westmeadows     9 Black St      3    h   582000.0      S   
13575  13575  Wheelers Hill   12 Strada Cr      4    h  1245000.0      S   
13576  13576   Williamstown  77 Merrett Dr      3    h  1031000.0     SP   
13577  13577   Williamstown    83 Power St      3    h  1170000.0      S   
13578  13578   Williamstown   96 Verdon St      4    h  2500000.0     PI   
13579  13579     Yarraville     6 Agnes St      4    h  1285000.0     SP   

             SellerG        Date  Distance  ...  Car  Landsize  BuildingArea  \
13573  hockingstuart  26/08/2017      14.7  ...  1.0     662.0         172.0   
13574            Red  26/08/2017      16.5  ...  2.0     256.0         126.0   
13575          Barry  26/08/2017      16.7  ...  2.0     652.0         126.0   
13576       Williams  26/08/2017       6.8  ...  2.0     333.0         133.0   
13577          Raine  26/08/2017       6.8  ...  4.0     436.0         126.0   
13578        Sweeney  26/08/2017       6.8  ...  5.0     866.0         157.0   
13579        Village  26/08/2017       6.3  ...  1.0     362.0         112.0   

       YearBuilt  CouncilArea  Lattitude  Longtitude  \
13573     1980.0          NaN  -37.89327   144.64789   
13574     1970.0          NaN  -37.67917   144.89390   
13575     1981.0          NaN  -37.90562   145.16761   
13576     1995.0          NaN  -37.85927   144.87904   
13577     1997.0          NaN  -37.85274   144.88738   
13578     1920.0          NaN  -37.85908   144.89299   
13579     1920.0          NaN  -37.81188   144.88449   

                       Regionname  Propertycount           Coordinates  
13573        Western Metropolitan        16166.0  -37.89327, 144.64789  
13574       Northern Metropolitan         2474.0   -37.67917, 144.8939  
13575  South-Eastern Metropolitan         7392.0  -37.90562, 145.16761  
13576        Western Metropolitan         6380.0  -37.85927, 144.87904  
13577        Western Metropolitan         6380.0  -37.85274, 144.88738  
13578        Western Metropolitan         6380.0  -37.85908, 144.89299  
13579        Western Metropolitan         6543.0  -37.81188, 144.88449  

[7 rows x 23 columns]

## Размерность таблицы 

Размер таблицы — количество строк и количество столбцов. Это можно сделать с помощью атрибута shape, который возвращает кортеж с количеством строк и столбцов:

In [49]:
melb_data.shape

(13580, 23)

Это говорит о том, что в исследуемой таблице содержится информация о 13 580 объектах недвижимости, и их описывают 23 признака.

## Получение информации о столбцах

Для того чтобы получить более детальную информацию о столбцах таблицы, можно использовать метод DataFrame info():

In [16]:
melb_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          13580 non-null  int64  
 1   Suburb         13580 non-null  object 
 2   Address        13580 non-null  object 
 3   Rooms          13580 non-null  int64  
 4   Type           13580 non-null  object 
 5   Price          13580 non-null  float64
 6   Method         13580 non-null  object 
 7   SellerG        13580 non-null  object 
 8   Date           13580 non-null  object 
 9   Distance       13580 non-null  float64
 10  Postcode       13580 non-null  int64  
 11  Bedroom        13580 non-null  int64  
 12  Bathroom       13580 non-null  int64  
 13  Car            13580 non-null  int64  
 14  Landsize       13580 non-null  float64
 15  BuildingArea   13580 non-null  float64
 16  YearBuilt      13580 non-null  int64  
 17  CouncilArea    12211 non-null  object 
 18  Lattit

Данный метод выводит:

* информацию об индексах;
* информацию об общем количестве столбцов;
* таблицу, в которой содержится информация об именах столбцов (Column), количестве непустых значений (Non-Null Count) в каждом столбце и типе данных столбца (Dtype), количестве столбцов, в которых используется определённый тип данных;
* количество оперативной памяти в мегабайтах, которое тратится на хранение данных.

Из вывода метода info() становится понятно, что в нашей таблице есть столбец CouncilArea с пропущенными значениями — количество непустых значений в столбце меньше, чем количество строк в таблице (12211 < 13580).

Пустыми, или пропущенными, значениями называются значения в ячейках таблицы, которые не заполнены по какой-либо причине, то есть на их месте стоит пустое место. В Pandas такие значения обозначаются символом NaN (Not-a-Number).

## Изменеие типа данных в столбце

Если присмотреться внимательнее к выводу метода info() при работе с данным DataFrame, а конкретнее — к типам данных столбцов, становится понятно, что некоторые признаки кодируются не совсем корректными типами данных.

Например, данные в столбцах, которые отражают количество, должны, по идее, выражаться целым числом (Car, Bedroom, Bathroom и Propertyсount), однако кодируются float64 — числом с плавающей запятой размером 64 бита.

Наконец, данные в столбце с годом постройки (YearBuilt) также представлены в формате чисел с плавающей точкой.

Чтобы исправить это, можно воспользоваться методом astype(), который позволяет преобразовать тип данных столбца:

In [10]:
melb_data['Car'] = melb_data['Car'].astype('int64')
melb_data['Bedroom'] = melb_data['Bedroom'].astype('int64')
melb_data['Bathroom'] = melb_data['Bathroom'].astype('int64')
melb_data['Propertycount'] = melb_data['Propertycount'].astype('int64')
melb_data['YearBuilt'] = melb_data['YearBuilt'].astype('int64')
melb_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          13580 non-null  int64  
 1   Suburb         13580 non-null  object 
 2   Address        13580 non-null  object 
 3   Rooms          13580 non-null  int64  
 4   Type           13580 non-null  object 
 5   Price          13580 non-null  float64
 6   Method         13580 non-null  object 
 7   SellerG        13580 non-null  object 
 8   Date           13580 non-null  object 
 9   Distance       13580 non-null  float64
 10  Postcode       13580 non-null  int64  
 11  Bedroom        13580 non-null  int64  
 12  Bathroom       13580 non-null  int64  
 13  Car            13580 non-null  int64  
 14  Landsize       13580 non-null  float64
 15  BuildingArea   13580 non-null  float64
 16  YearBuilt      13580 non-null  int64  
 17  CouncilArea    12211 non-null  object 
 18  Lattit

В данном коде мы при помощи метода astype() последовательно переопределяем столбцы на них же самих, только с изменённым типом данных: int64 — целочисленное число размером 64 бита. Дополнительно о типах данных здесь https://dfedorov.spb.ru/pandas/%D0%9E%D0%B1%D0%B7%D0%BE%D1%80%20%D1%82%D0%B8%D0%BF%D0%BE%D0%B2%20%D0%B4%D0%B0%D0%BD%D0%BD%D1%8B%D1%85%20pandas.html

## Получение описательной статистики

Часто при работе с таблицей нужно быстро посмотреть на основные статистические свойства её столбцов. Для этого можно воспользоваться методом DataFrame describe().

По умолчанию метод работает с числовыми (int64 и float64) столбцами и показывает число непустых значений (count), среднее (mean), стандартное отклонение (std), минимальное значение (min),  квантили уровней 0.25, 0.5 (медиана) и 0.75 (25%, 50%, 75%) и максимальное значение (max) для каждого столбца исходной таблицы.

Bыведем на экран значение статистических параметров только для столбцов Distance (расстояние от объекта недвижимости до центра Мельбурна), BuildingArea (площадь здания) и Price (цена объекта):

In [11]:
melb_data.describe().loc[:, ['Distance', 'BuildingArea' , 'Price']]


Distance  BuildingArea         Price
count  13580.000000  13580.000000  1.358000e+04
mean      10.137776    139.633972  1.075684e+06
std        5.868725    392.217403  6.393107e+05
min        0.000000      0.000000  8.500000e+04
25%        6.100000    122.000000  6.500000e+05
50%        9.200000    126.000000  9.030000e+05
75%       13.000000    129.940000  1.330000e+06
max       48.100000  44515.000000  9.000000e+06

Какие интересные выводы мы можем сделать, исходя из построенной таблицы?

 Столбец **Distance** (расстояние до центра города)

* В нашем наборе данных представлены объекты, удалённые от центра города на расстояние до 48 километров.

 Столбец **BuildingArea** (площадь здания)

* В нашей таблице представлены участки, на которых отсутствует само здание (его площадь равна 0), то есть владельцы продавали участок земли без строений на нём.
* Большая часть объектов (75%) имеет площадь здания до 130 квадратных метров, однако в нашей таблице есть настоящий дворец (судя по его размерам) с площадью в 44 515 квадратных метров.
 
 Столбец **Price** (цена)

* Диапазон цен на недвижимость в Мельбурне варьируется от 85 тысяч до 9 миллионов австралийских долларов.

На самом деле метод describe() можно применять не только к числовым признакам. С помощью параметра include можно указать тип данных, для которого нужно вывести описательную информацию.

Например, для типа данных object метод describe() возвращает DataFrame, в котором указаны:

* количество непустых строк (count);
* количество уникальных значений (unique);
* самое частое значение — мода —  (top);
* частота — объём использования — этого значения (freq) для каждого столбца типа object исходной таблицы.

In [12]:
melb_data.describe(include=['object'])

Suburb           Address   Type Method SellerG        Date  \
count       13580             13580  13580  13580   13580       13580   
unique        314             13378      3      5     268          58   
top     Reservoir  36 Aberfeldie St      h      S  Nelson  27/05/2017   
freq          359                 3   9449   9022    1565         473   

       CouncilArea             Regionname         Coordinates  
count        12211                  13580               13580  
unique          33                      8               13097  
top       Moreland  Southern Metropolitan  -37.8361, 144.9966  
freq          1163                   4695                  12

Рассострим, о чем свидетельствуют данные

 Столбец **Suburb** (пригород)

* Наибольшее количество проданных объектов (359) находилось в пригороде Reservoir.

 Столбец **Type** (тип здания)

* Самый популярный тип дома — h (дом, вилла, коттедж).

 Столбец **SellerG** (риелтор)

* В наших данных нам известно о 268 различных риэлторских компаниях, однако самой главной «акулой» в этом бизнесе является компания Nelson — они продали 1 565 различных домов.

 Столбец **Date** (дата продажи)

* В нашей таблице содержится информация за 58 дней, при этом наибольшее число продаж (473) пришлось на 27 мая 2017 года.

## Получение частоты уникальных значений в столбце

Для того чтобы определить, сколько раз в столбце повторяется каждый из вариантов значений (т.е. найти частоту для каждого уникального знания), используется метод value_counts().

Данный метод возвращает объект Series, в котором в качестве индексов выступают уникальные категории столбца, а значениями — соответствующая им частота.

In [13]:
# Рассмотрим работу value_counts() на примере столбца с названиями районов:
melb_data['Regionname'].value_counts()

Southern Metropolitan         4695
Northern Metropolitan         3890
Western Metropolitan          2948
Eastern Metropolitan          1471
South-Eastern Metropolitan     450
Eastern Victoria                53
Northern Victoria               41
Western Victoria                32
Name: Regionname, dtype: int64

Чтобы сделать вывод более интерпретируемым и понятным, можно воспользоваться параметром normalize. При установке значения этого параметра на True результат будет представляться в виде доли (относительной частоты):

In [14]:
melb_data['Regionname'].value_counts(normalize=True)


Southern Metropolitan         0.345729
Northern Metropolitan         0.286451
Western Metropolitan          0.217084
Eastern Metropolitan          0.108321
South-Eastern Metropolitan    0.033137
Eastern Victoria              0.003903
Northern Victoria             0.003019
Western Victoria              0.002356
Name: Regionname, dtype: float64

Из результатов мы можем сделать вывод:  
* наименьшее количество объектов (менее 1%) было продано в районах Victoria, а наибольшее — в районах Metropolitan
* в столбце Regionname восемь уникальных значений (unique), что даёт право сделать вывод, что в Мельбурне и его пригородах как минимум восемь различных районов.
* количество уникальных значений (unique) в столбце Address больше, чем количество уникальных значений в столбце Coordinates. Это значит, что некоторые различные адреса помечены одинаковыми географическими координатами.
* наиболее часто встречающееся (top) значение в столбце Suburb является Reservoir. Частота (freq) его использований — 359 раз.

In [15]:
data = pd.DataFrame([[0,1], [1, 0], [1, 1]], columns=['А', 'B'])
print(data.shape)

(3, 2)


In [18]:
# Выведем процент от общего количества домов - таунхаусы (тип объекта — t)?
melb_data['Type'].value_counts(normalize=True)

h    0.695803
u    0.222165
t    0.082032
Name: Type, dtype: float64

## Статистические методы. Агрегирующие методы

С помощью метода describe() можно выводить информацию о статистических показателях. Однако этот метод становится не очень удобным, когда необходимо найти только один статистический параметр, например только среднюю цену,  и использовать их в дальнейшем коде. Поэтому в Pandas предусмотрены инструменты быстрого вычисления показателей с помощью агрегирующих методов. 

Агрегирующим в Pandas называется метод, который для каждого столбца возвращает только одно значение — показатель (например, вычисление медианы, максимума, среднего и так далее).

Ниже приведена таблица основных агрегирующих методов:

  МЕТОД            СТАТИСТИЧЕСКИЙ ПАРАМЕТР
* .count()	    --  Количество непустых значений
* .mean()	      --  Среднее значение
* .min()	      --  Минимальное значение
* .max()	      --  Максимальное значение
* .var()	      --  Дисперсия
* .std()	      --  Стандартное отклонение
* .sum()	      --  Сумма
* .quantile(x)	--  Квантиль уровня x
* .nunique()	  --  Число уникальных значений

Если один из этих методов применить ко всему DataFrame, то в результате его работы будет получен объект типа Series, в котором в качестве индексов будут выступать наименования столбцов, а в качестве значений — статистический показатель. В случае применения метода к отдельному столбцу результатом вычислений станет число.

В каждый метод можно передать некоторые параметры, среди которых:

*  axis  — определяет, подсчитывать параметр по строкам или по столбцам;
*  numeric_only — определяет, вычислять параметры только по числовым столбцам/строкам или нет (True/False).

In [3]:
# Вычислим среднюю цену на объекты недвижимости:

print(melb_data['Price'].mean())

# Найдём максимальное количество парковочных мест:

print(melb_data['Car'].max())

1075684.079455081
10.0


А теперь представим, что риэлторская ставка для всех компаний за продажу недвижимости составляет 12%.

In [4]:
# Найдём общую прибыльность риэлторского бизнеса в Мельбурне. 
# Результат округлим до сотых:

rate = 0.12
income = melb_data['Price'].sum() * rate
print('Total income of real estate agencies:', round(income, 2))

Total income of real estate agencies: 1752934775.88


Найдём, насколько медианная площадь территории отличается от её среднего значения. Вычислим модуль разницы между медианой и средним и разделим результат на среднее, чтобы получить отклонение в долях:

In [5]:
landsize_median = melb_data['Landsize'].median() 
landsize_mean =  melb_data['Landsize'].mean()
print(abs(landsize_median - landsize_mean)/landsize_mean)

0.21205713983546193


В результате получаем долю отклонения медианы от среднего значения. Умножив результат на 100, получим его в процентах. Отклонение медианы от среднего значения на 21% является довольно большим, и это повод задуматься над тем, чтобы исследовать признак на наличие аномалий. 

## Статистические методы. Модальное значение

Отдельный интерес представляет статический показатель моды — самого распространённого значения в столбце. Он вычисляется с помощью метода mode().

Модальных значений может быть несколько, то есть несколько значений могут встречаться одинаковое количество раз. Поэтому метод mode(), в отличие от агрегирующих методов, возвращает не одно число, а серию.

In [6]:
# Вычислим, какое число комнат чаще всего представлено на рынке недвижимости:

print(melb_data['Rooms'].mode())

0    3
Name: Rooms, dtype: int64


Метод mode() может быть использован не только с числовыми столбцами, но и со столбцами типа object. 

In [7]:
# Найдем наиболее распространённое название района:

melb_data['Regionname'].mode()

0    Southern Metropolitan
Name: Regionname, dtype: object

Метод std() вычисляет стандартное отклонение и работает только с числовыми столбцами.

In [15]:
melb_data['Landsize'].std()

3990.6692411090316

Данные в столбце Landsize относятся к числовому типу, следовательно, метод не выдаст ошибки.

In [16]:
melb_data['Price'].std()

639310.724296018

Данные в столбце Price относятся к числовому типу, следовательно, метод не выдаст ошибки.

In [ ]:
melb_data['Date'].std()

Данные в столбце Date относятся к строковому типу, следовательно, метод выдаст ошибку.

In [ ]:
melb_data['Type'].std()

Данные в столбце Type относятся к строковому типу, следовательно, метод выдаст ошибку.

In [21]:
# Mаксимальное количество домов на продажу в районе (Propertycount)
print(melb_data['Propertycount'].max())

21650.0


In [22]:
# Cтандартное отклонение (разброс) расстояния 
# от центра города до объекта недвижимости
melb_data['Distance'].std()

5.868724943071711

In [25]:
# Тоже стандартное отклонение (разброс) расстояния 
# от центра города до объекта недвижимости
# если ответ округляем до 2 знаков после запятой
distance = melb_data['Distance'].std()
print(round(distance, 2))

5.87


In [23]:
# Oтклонение (в процентах) медианного значения 
# площади здания от его среднего значения
building_median = melb_data['BuildingArea'].median() 
building_mean =  melb_data['BuildingArea'].mean()
deviance = abs(building_median - building_mean)/building_mean
print(round(deviance * 100, 2))

9.76


In [30]:
# Задан ряд чисел [1, 2, 4, 5, 3, 5, 1, 5, 6]. Чему равна мода в данном ряду?
num = [1, 2, 4, 5, 3, 5, 1, 5, 6]
mode_ = max((num.count(item), item) for item in set(num))[1]
print(f'Вычисление моды: {mode_}')
# Мода — наиболее часто встречающееся значение в ряду

Вычисление моды: 5


In [28]:
# Задан ряд чисел [2, 3, 2, 8, 12]. Чему равна мода в данном ряду?
u = [2, 3, 2, 8, 12]
mode_ = max((u.count(item), item) for item in set(u))[1]
print(f'Вычисление моды: {mode_}')

Вычисление моды: 2


In [33]:
# Сколько спален чаще всего встречается в домах в Мельбурне?
bed = melb_data['Bedroom'].mode()

0    3.0
Name: Bedroom, dtype: float64


## Фильтрация данных в DataFrame

Часто возникает необходимость исследовать определённую группу объектов по какому-то условию, например найти здания с ценой меньше 1 миллиона или выделить из всей таблицы помещения с двумя комнатами.

Такие задачи называются задачами фильтрации.

Под фильтрацией в DataFrame подразумевается получение новой таблицы путём вырезания строк, не удовлетворяющих поставленному условию. 

Разберём классический способ фильтрации в DataFrame — фильтрацию с помощью масок.

Маской называется Series, которая состоит из булевых значений, при этом значения True соответствуют тем индексам, для которых заданное условие выполняется, в противном случае ставится значение False (например, цена > 2 млн).

In [ ]:
# Создадим маску и положим её в переменную с именем mask:

mask = melb_data['Price'] > 2000000
display(mask)

Для фильтрации нужно просто подставить переменную mask в индексацию DataFrame. Маска показывает, какие строки нужно оставлять в результирующем наборе, а какие — убирать:

In [ ]:
# Bыведем первые пять строк отфильтрованной таблицы
display(melb_data[mask].head())

В результате выполнения фильтрации возвращается новый DataFrame, полученный из исходного, при этом исходная таблица melb_data остаётся без изменений.

Также вовсе не обязательно заносить маску в отдельную переменную — можно сразу вставлять условие в операцию индексации DataFrame, например:

In [35]:
melb_data[melb_data['Price'] > 2000000]

index          Suburb               Address  Rooms Type      Price  \
80        80     Albert Park  112 Beaconsfield Pde      3    h  2850000.0   
85        85     Albert Park     104 Richardson St      4    h  2300000.0   
88        88     Albert Park        29 Faussett St      2    h  2120000.0   
92        92     Albert Park           2 Dundas Pl      3    h  2615000.0   
93        93     Albert Park          23 Finlay St      5    h  2100000.0   
...      ...             ...                   ...    ...  ...        ...   
13521  13521  Port Melbourne          44 Garton St      4    t  2455000.0   
13523  13523         Prahran        69 Greville St      4    h  2668000.0   
13553  13553    Surrey Hills          20 Albert Cr      4    h  2720000.0   
13555  13555    Surrey Hills              3 Oak St      4    h  3100000.0   
13578  13578    Williamstown          96 Verdon St      4    h  2500000.0   

      Method   SellerG        Date  Distance  ...  Car  Landsize  \
80        PI    Buxton   4/03/2017       3.3  ...  0.0     211.0   
85         S  Marshall   7/05/2016       3.3  ...  1.0     153.0   
88         S      Greg  10/09/2016       3.3  ...  1.0     199.0   
92         S    Cayzer  10/12/2016       3.3  ...  1.0     177.0   
93         S      Greg  10/12/2016       3.3  ...  1.0     237.0   
...      ...       ...         ...       ...  ...  ...       ...   
13521     SP  Marshall  26/08/2017       3.5  ...  2.0     123.0   
13523      S    Biggin  26/08/2017       4.6  ...  2.0     383.0   
13553      S       Kay  26/08/2017      10.2  ...  2.0    1005.0   
13555     VB  Marshall  26/08/2017      10.2  ...  3.0     832.0   
13578     PI   Sweeney  26/08/2017       6.8  ...  5.0     866.0   

       BuildingArea  YearBuilt   CouncilArea  Lattitude  Longtitude  \
80            198.0     1890.0  Port Phillip  -37.84810   144.94990   
85            180.0     1880.0  Port Phillip  -37.84470   144.95230   
88            107.0     1900.0  Port Phillip  -37.84220   144.95540   
92            181.0     1880.0  Port Phillip  -37.84150   144.95850   
93            126.0     1970.0  Port Phillip  -37.84360   144.95570   
...             ...        ...           ...        ...         ...   
13521           0.0     2010.0           NaN  -37.83349   144.94840   
13523         126.0     1970.0           NaN  -37.84879   144.98882   
13553         126.0     1920.0           NaN  -37.82421   145.10352   
13555         126.0     1970.0           NaN  -37.83564   145.10919   
13578         157.0     1920.0           NaN  -37.85908   144.89299   

                  Regionname  Propertycount           Coordinates  
80     Southern Metropolitan         3280.0    -37.8481, 144.9499  
85     Southern Metropolitan         3280.0    -37.8447, 144.9523  
88     Southern Metropolitan         3280.0    -37.8422, 144.9554  
92     Southern Metropolitan         3280.0    -37.8415, 144.9585  
93     Southern Metropolitan         3280.0    -37.8436, 144.9557  
...                      ...            ...                   ...  
13521  Southern Metropolitan         8648.0   -37.83349, 144.9484  
13523  Southern Metropolitan         7717.0  -37.84879, 144.98882  
13553  Southern Metropolitan         5457.0  -37.82421, 145.10352  
13555  Southern Metropolitan         5457.0  -37.83564, 145.10919  
13578   Western Metropolitan         6380.0  -37.85908, 144.89299  

[1007 rows x 23 columns]

Найдём количество зданий с тремя комнатами. Для этого отфильтруем таблицу по условию: обратимся к результирующей таблице по столбцу Rooms и найдём число строк в ней с помощью атрибута shape:

In [36]:
melb_data[melb_data['Rooms'] == 3].shape[0]

5881

Условия можно комбинировать, используя операторы & (логическое И) и | (логическое ИЛИ). Условия при этом заключаются в скобки.

In [37]:
# Hайдём число трёхкомнатных домов с ценой менее 300 тысяч:
melb_data[(melb_data['Rooms'] == 3) & (melb_data['Price'] < 300000)].shape[0]

3

Таких зданий оказалось всего три. Немного «ослабим» условие: теперь нас будут интересовать дома с ценой менее 300 тысяч, у которых либо число комнат равно 3 либо площадь домов более 100 квадратных метров:

In [38]:
melb_data[((melb_data['Rooms'] == 3) | (melb_data['BuildingArea'] > 100)) & (melb_data['Price'] < 300000)].shape[0]

68

или с использованием **mask**

In [5]:
mask1 = (melb_data['Rooms'] == 3)
mask2 = (melb_data['BuildingArea'] > 100)
mask3 = (melb_data['Price'] < 300000)
melb_data[(mask1 | mask2) & mask3].shape[0]
# так код выглядит более элегантно


68

Важно, что использование привычных операторов and и or будет неверным и приведёт к ошибке, так как они выполняют логические операции между двумя булевыми числами. В нашем случае слева и справа от оператора стоят маски (объекты Series), для которых логическую операцию надо совершить поэлементно, а операторы and и or для такого не предназначены.

Фильтрацию часто сочетают со статистическими методами. Давайте найдём максимальное количество комнат в таунхаусах. Так как в результате фильтрации получается DataFrame, то обратимся к нему по столбцу Rooms и найдём максимальное значение:

In [39]:
melb_data[melb_data['Type'] == 't']['Rooms'].max()


5

А теперь более сложный трюк: найдём медианную площадь здания у объектов, чья цена выше средней. Для того чтобы оградить наш код от нагромождений, предварительно создадим переменную со средней ценой:

In [40]:
mean_price = melb_data['Price'].mean()
melb_data[melb_data['Price'] > mean_price]['BuildingArea'].median()

126.0

Фильтрация находит применение в очистке данных.

Под очисткой данных понимается удаление из данных аномальных значений (выбросов), пропусков и данных, которые не несут информацию.

Причин появления таких дефектов в данных множество: неправильная передача по сети, неверная выгрузка из базы, человеческий фактор и т. д.

→ Проблема аномальных значений состоит в том, что если мы обучаем модель на данных, содержащих выбросы, то мы имеем немалый шанс получить менее верный прогноз, чего нам, конечно же, не хотелось бы.

In [41]:
# Число объектов недвижимости из таблицы melb_data без ванных комнат
melb_data[melb_data['Bathroom'] == 0].shape[0]

34

In [43]:
# Число объектов недвижимости, 
# проданных риелторской компанией Nelson по цене больше 3 миллионов
melb_data[(melb_data['SellerG'] == 'Nelson') & (melb_data['Price'] > 3000000)].shape[0]
# melb_data[(melb_data['SellerG'] == 'Nelson') & (melb_data['Price'] > 3e6)].shape[0]

5

In [45]:
# Mинимальная стоимость участка без здания (площадь здания равна 0) 
melb_data[(melb_data['BuildingArea'] == 0)]['Price'].min()

412500.0

Какова средняя цена объектов недвижимости в таблице melb_data с ценой менее одного миллиона, в которых либо количество комнат больше пяти, либо здание моложе 2015 года?

In [47]:
round(melb_data[(melb_data['Price']<1e6) & ((melb_data['Rooms'] > 5) | (melb_data['YearBuilt'] > 2015))]['Price'].mean())

769239

или с использованием **mask**

In [4]:
mask1 = (melb_data['Price']<1e6)
mask2 = (melb_data['Rooms'] > 5) 
mask3 = (melb_data['YearBuilt'] > 2015)
round(melb_data[(mask1 & (mask2 | mask3))]['Price'].mean())

769239

In [52]:
# Pайон Мельбурна чаще всего продаются виллы и коттеджи (тип здания — h) 
# с ценой меньше трёх миллионов
melb_data[(melb_data['Type'] == 'h') & (melb_data['Price'] < 3e6)]['Regionname'].mode()

0    Northern Metropolitan
Name: Regionname, dtype: object

С использованием **Mask**

In [55]:
mask1 = (melb_data['Type'] == 'h')
mask2 = (melb_data['Price'] < 3e6)
melb_data[(mask1 & mask2)]['Regionname'].mode()

0    Northern Metropolitan
Name: Regionname, dtype: object

In [7]:
# Импортируем библиотеку и DataFrame
import pandas as pd
students = pd.read_csv('data/students_performance.csv', sep=',')

# Выведем первые пять строк нашей таблицы:
display(students.head())

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

Информация о столбцах таблицы:

* gender — пол;
* race/ethnicity — раса/этническая принадлежность;
* parental level of education — уровень образования родителей;
* lunch — какие обеды получал студент во время обучения (standard — платный, free/reduced — бесплатный);
* test preparation course — посещал ли студент курсы подготовки к экзаменам (none — не посещал, completed — посещал);
* math score, reading score, writing score — баллы по математике, чтению и письму по сто балльной шкале.

Узнаем число студентов в таблице

In [8]:
display(students.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


None

In [12]:
# балл по письму у студента под индексом 155
students.loc[155, 'writing score']

88

In [13]:
# Узнаем средний балл
students.mean(axis=0)

C:\Users\Гарик\AppData\Local\Temp\ipykernel_4692\22535865.py:1: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  students.mean(axis=0)


math score       66.089
reading score    69.169
writing score    68.054
dtype: float64

In [14]:
# Какая расовая группа является самой крупной в учебном заведении?
students['race/ethnicity'].mode()

0    group C
Name: race/ethnicity, dtype: object

In [17]:
# Каков средний балл по чтению у студентов, 
# которые посещали курсы подготовки к экзаменам?
students[(students['test preparation course'] == 'completed')]['reading score'].mean()

# Округлим ответ до целого числа
# round(students[(students['test preparation course'] == 'completed')]['reading score'].mean())

74

In [23]:
# Число студентов, которые получили 0 баллов по математике
mask = students['math score'] == 0
display(students[mask].info())


<class 'pandas.core.frame.DataFrame'>
Int64Index: 1 entries, 59 to 59
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1 non-null      object
 1   race/ethnicity               1 non-null      object
 2   parental level of education  1 non-null      object
 3   lunch                        1 non-null      object
 4   test preparation course      1 non-null      object
 5   math score                   1 non-null      int64 
 6   reading score                1 non-null      int64 
 7   writing score                1 non-null      int64 
dtypes: int64(3), object(5)
memory usage: 72.0+ bytes


None

In [26]:
# Фильтруем студентов с оплачиваемым питанием
mask1 = (students['lunch'] == 'standard')

# Фильтруем студентов с льготным питанием
mask2 = (students['lunch'] == 'free/reduced') 

# Cредний балл по математике у студентов с оплачиваемым питанием 
print(round(students[(mask1)]['math score'].mean()))

# Cредний балл по математике у студентов с льготным питанием
print(round(students[(mask2)]['math score'].mean()))


70


59

Каков процент студентов, родители которых имеют высшее образование уровня бакалавриата (bachelor's degree)?

In [27]:
students["parental level of education"].value_counts(normalize=True)

some college          0.226
associate's degree    0.222
high school           0.196
some high school      0.179
bachelor's degree     0.118
master's degree       0.059
Name: parental level of education, dtype: float64

Насколько медианный балл по письму у студентов в расовой группе А отличается от среднего балла по письму у студентов в расовой группе C?

In [28]:
a = students[students['race/ethnicity'] == 'group A']['writing score'].median()
b = students[students['race/ethnicity'] == 'group C']['writing score'].mean()
print(round(abs(a - b)))

6
